File: src/services/cache.js
const NodeCache = require('node-cache');
const crypto = require('crypto');

class CacheService {
  constructor(config) {
    // TTL is in seconds for node-cache
    this.cache = new NodeCache({ stdTTL: config.CACHE_TTL });
  }

  get(key) {
    return this.cache.get(key);
  }

  set(key, value) {
    this.cache.set(key, value);
  }

  generateKey(query, numResults, mode, engines, timeRange = "") {
    const data = `${query}|${numResults}|${mode}|${engines.join(',')}|${timeRange}`;
    const hashVal = crypto.createHash('md5').update(data).digest('hex');
    return `search:${hashVal}`;
  }
}

module.exports = CacheService;

File: src/services/cache.py
from cachetools import TTLCache
from config.settings import CACHE_TTL

cache = TTLCache(maxsize=1000, ttl=CACHE_TTL)

def get(key: str):
    return cache.get(key)

def set(key: str, value):
    cache[key] = value

def generate_key(query: str, num_results: int, mode: str, engines: list, time_range: str = "") -> str:
    data = f"{query}|{num_results}|{mode}|{','.join(engines or [])}|{time_range}"
    hash_val = hash(data)
    return f"search:{abs(hash_val)}"
